pkg/costattribution/tracker.go

```diff
@@ -0,0 +1,371 @@
+}
+
+func (t *Tracker) updateCounters(lbls labels.Labels, ts int64, activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement float64, reason *string, createIfDoesNotExist bool) {
+	extractValues := func() []string {
+		labelValues := make([]string, len(t.labels))
+		for idx, cal := range t.labels {
+			labelValues[idx] = lbls.Get(cal)
+			if labelValues[idx] == "" {
+				labelValues[idx] = missingValue
+			}
+		}
+		return labelValues
+	}
+	t.updateCountersCommon(extractValues, ts, activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement, reason, createIfDoesNotExist)
+}
+
+func (t *Tracker) updateCountersCommon(
+	extractValues func() []string,
+	ts int64,
+	activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement float64,
+	reason *string,
+	createIfDoesNotExist bool,
+) {
+	// Extract label values
+	labelValues := extractValues()
+
+	// Reuse buffer from pool for building the observation key
+	buf := bufferPool.Get().(*bytes.Buffer)
+	buf.Reset()
+	defer bufferPool.Put(buf)
+
+	// Construct the observation key by joining label values
+	for i, value := range labelValues {
+		if i > 0 {
+			buf.WriteRune(sep)
+		}
+		buf.WriteString(value)
+	}
+
+	// Lock access to the observation map
+	t.observedMtx.Lock()
+	defer t.observedMtx.Unlock()
+
+	// Update observations and state
+	t.updateObservations(buf.Bytes(), ts, activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement, reason, createIfDoesNotExist)
+	t.updateState(ts, activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement)
+}
+
+// updateObservations updates or creates a new observation in the 'observed' map.
+func (t *Tracker) updateObservations(key []byte, ts int64, activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement float64, reason *string, createIfDoesNotExist bool) {
+	if o, known := t.observed[string(key)]; known && o.lastUpdate != nil {
+		o.lastUpdate.Store(ts)
+		if activeSeriesIncrement != 0 {
+			o.activeSerie.Add(activeSeriesIncrement)
+		}
+		if receivedSampleIncrement > 0 {
+			o.receivedSample.Add(receivedSampleIncrement)
+		}
+		if discardedSampleIncrement > 0 && reason != nil {
+			o.discardedSampleMtx.Lock()
+			if v, ok := o.discardedSample[*reason]; ok {
+				v.Add(discardedSampleIncrement)
+				o.discardedSample[*reason] = v
+			} else {
+				o.discardedSample[*reason] = *atomic.NewFloat64(discardedSampleIncrement)
+			}
+			o.discardedSampleMtx.Unlock()
+		}
+	} else if len(t.observed) < t.maxCardinality*2 && createIfDoesNotExist {
+		// When createIfDoesNotExist is false, it means that the method is called from DecrementActiveSeries, when key doesn't exist we should ignore the call
+		// Otherwise create a new observation for the key
+		t.createNewObservation(key, ts, activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement, reason)
+	}
```

Comment:
I would move this to the beginning of the method to un-indent the most common path.
i.e.
```go
o, known := t.observed[string(key)]
if !known {
	if len(t.observed) < t.maxCardinality*2 && createIfDoesNotExist {
		// When createIfDoesNotExist is false, it means that the method is called from DecrementActiveSeries, when key doesn't exist we should ignore the call
		// Otherwise create a new observation for the key
		t.createNewObservation(key, ts, activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement, reason)
	}
	return
}
// ...
```

Reply:
addressed in 9c4c2dfa32e4961b412e0d91eb391e9184b80c03